### ISSUE TITLE
Email users once bulk upload is complete

### ISSUE THREAD

**Issue by emmastephenson** [CDCgov/prime-simplereport#4440]
### Background
We need a way to notify users that their patient uploads are complete. Doing this in-app (via a notification toast) would be tricky, but we have a pattern established elsewhere to email users directly upon completion of a backend action.

### Action Requested
Once a patient bulk upload has finished processing, email the user who initiated the upload to notify them that the upload has finished. We should be able to fetch the user's email directly out of something like [ApiUserContextHolder](https://github.com/CDCgov/prime-simplereport/blob/1ccbd4b2a982442696abee1b0d0bbc17553af956/backend/src/main/java/gov/cdc/usds/simplereport/api/ApiUserContextHolder.java) or one of our other permissioning helpers, but if not the email can be passed directly along with the CSV upload request.

We already have an [EmailService](https://github.com/CDCgov/prime-simplereport/blob/1ccbd4b2a982442696abee1b0d0bbc17553af956/backend/src/main/java/gov/cdc/usds/simplereport/service/email/EmailService.java#L21-L21) that should be sufficient to trigger these emails. We'll also need to create a new email template in SendGrid with the content of the new email and link it to our site using the [EmailProviderTemplate](https://github.com/CDCgov/prime-simplereport/blob/1ccbd4b2a982442696abee1b0d0bbc17553af956/backend/src/main/java/gov/cdc/usds/simplereport/service/email/EmailProviderTemplate.java#L3-L3).

You can see an example of the email trigger in the [IdentityVerificationController](https://github.com/CDCgov/prime-simplereport/blob/1ccbd4b2a982442696abee1b0d0bbc17553af956/backend/src/main/java/gov/cdc/usds/simplereport/api/accountrequest/IdentityVerificationController.java#L189-L189), where we email users next steps once they've failed the Experian id verification. (Worth noting that it's preferred to use the SendGrid dynamic template for sending emails, rather than creating the HTML directly in SimpleReport as is done when we [send emails to SR support](https://github.com/CDCgov/prime-simplereport/blob/1ccbd4b2a982442696abee1b0d0bbc17553af956/backend/src/main/java/gov/cdc/usds/simplereport/api/accountrequest/IdentityVerificationController.java#L183-L183)).

### Acceptance criteria
The user who initiates a patient bulk upload is emailed when all the patients have been processed, loaded into the database, and are available to view in SimpleReport.

**Comment by georgehager** [comment]
@emmastephenson will this actually be resolved in the work you're currently doing? I think I'll probably at least need to wait until that work is done, is that right?

**Comment by lauraykerr** [comment]
@kenieh What do you think about this copy for the email notification?

> Greetings from SimpleReport, 
>
> Your patient bulk upload is complete. 
>
> You can now view and report test results for the new patients on SimpleReport. 
>
> [Button] See patients

**Comment by georgehager** [comment]
@kenieh @lauraykerr I've drafted a template in SendGrid, will update as needed once Kenny is able to take a look! One question - What would you like to see below the email body? Our template in SendGridseems to have `Unsubscribe - Unsubscribe Preferences` as a default, but I don't think that's being actively used and not sure how we would be set up those links anyway. I do see SimpleReport terms of service in other emails though.

**Current template**
![Screen Shot 2022-11-18 at 12 49 22 PM](https://user-images.githubusercontent.com/89797785/202769986-aab37311-bab7-4490-acc3-ded6cbad7d14.png)
**Example of what existing emails look like**
![Screen Shot 2022-11-18 at 12 51 32 PM](https://user-images.githubusercontent.com/89797785/202769999-2acb7943-efe2-4c4c-9d5d-bf24a5d6ca63.png)
![Screen Shot 2022-11-18 at 12 53 14 PM](https://user-images.githubusercontent.com/89797785/202770317-7e785c71-b163-4fed-a678-4aca8a245be1.png)

**Comment by georgehager** [comment]
@emmastephenson @kenieh @lauraykerr Where do you want the button link to go? Is it sufficient to just have `https://simplereport.gov` or would you prefer a unique url with the uploader's organization, etc?

**Comment by emmastephenson** [comment]
If we could get a unique URL that points to the "Patients" tab for the organization/facility the user uploaded to that would be nice, but I'm not convinced that URL would even be respected if the person is logged out of SimpleReport (I think it would redirect to the login page, then the facility selector, then the "Conduct Tests" page as usual.) But maybe still worth including in case folks are logged in?

On the unsubscribe front - I think it's actually mandated to include an unsubscribe link in most emails these days; I know that's at least the case with our text messages where we need to include a clear opt-out option. I believe that's all handled by SendGrid though, so I don't think there's any configuration/links we'd need to set up.

**Comment by georgehager** [comment]
>If we could get a unique URL that points to the "Patients" tab for the organization/facility the user uploaded to that would be nice, but I'm not convinced that URL would even be respected if the person is logged out of SimpleReport (I think it would redirect to the login page, then the facility selector, then the "Conduct Tests" page as usual.) But maybe still worth including in case folks are logged in?

That was my concern, too. But we can add it in case they are logged in, if Kenny and Laura agree.

> On the unsubscribe front - I think it's actually mandated to include an unsubscribe link in most emails these days; I know that's at least the case with our text messages where we need to include a clear opt-out option. I believe that's all handled by SendGrid though, so I don't think there's any configuration/links we'd need to set up.

awesome, thanks @emmastephenson! I wonder if we need to add that to some of our email templates, I notice we're not including the `Unsubscribe` in some of them.

**Comment by emmastephenson** [comment]
> awesome, thanks @emmastephenson! I wonder if we need to add that to some of our email templates, I notice we're not including the `Unsubscribe` in some of them.

Yeah I certainly don't think it would hurt, that's something we regularly get dinged for on Twilio (you eventually run the risk of service providers thinking you're spam otherwise). Do you mind writing up a separate ticket for that?

**Comment by georgehager** [comment]
> Yeah I certainly don't think it would hurt, that's something we regularly get dinged for on Twilio (you eventually run the risk of service providers thinking you're spam otherwise). Do you mind writing up a separate ticket for that?

Yep will do!

**Comment by kenieh** [comment]
Short and simple, I like! For the copy, could we condense it a bit so there isn't so much white space? All I did was make the body one paragraph rather than two paragraphs.

> Greetings from SimpleReport,
> 
> Your patient bulk upload is complete. You can now view and report test results for the new patients on SimpleReport.
> 
> [Button] See patients

**Comment by kenieh** [comment]
As for what goes below the email content, I'd default to what we have showing in other emails. 
i.e. [note I added the support email]

> Still have questions? Visit our Support Page or email us at support@simplereport.cdc.gov.
> 
> Thanks for choosing SimpleReport.
> 
> SimpleReport terms of service.

**Comment by kenieh** [comment]
How much control do we have over the email design? that would also help me understand our technical constraints with emails.

**Comment by emmastephenson** [comment]
@kenieh this is the one instance where we have a lot of control over the email 😄 It's just an HTML template, so we can do whatever we want within the limits of raw HTML. The email templates are in SendGrid if you want to take a look; we also have some templates for Okta in this static site folder:https://github.com/CDCgov/prime-simplereport-site/tree/feb30b2f700ae18e248ba128387610c4aefbc745/assets/user_emails

**Comment by georgehager** [comment]
@kenieh how does this look? I made the text underneath the button smaller ("Still have questions...") because it otherwise would be on two lines.

![Screen Shot 2022-11-21 at 10 30 05 AM](https://user-images.githubusercontent.com/89797785/203094091-7c76c650-24d3-4181-9f17-784eef64e4ec.png)

**Comment by lauraykerr** [comment]
This looks great! @georgehager Is there a reason "see patients" is written again below the button? Preference would be to take it out unless we need it there for accessibility or some other reason. 

I think including the unique URL for patient tab is fine, even for those not logged in.

**Comment by georgehager** [comment]
> This looks great! @georgehager Is there a reason "see patients" is written again below the button? Preference would be to take it out unless we need it there for accessibility or some other reason.

> I think including the unique URL for patient tab is fine, even for those not logged in.

Hey @lauraykerr, I can take that "see patients" text out - it was in another template like that but easy to remove.

I've updated to include the unique url!

**Comment by georgehager** [comment]
@lauraykerr @kenieh another note, after consultation with @emmastephenson, I've removed the "Unsubscribe - Unsubscribe Preferences" section, as this was setup for okta but does not work in sendgrid without significant extra configuration. Since we aren't sending marketing emails, it should be fine to leave it out. (We leave it out on our other emails as well.)

FInal draft (pending your approval!):

![Screen Shot 2022-11-21 at 4 02 36 PM](https://user-images.githubusercontent.com/89797785/203157937-34686cc3-a61a-46cb-8ce8-f53e9d23b71a.png)

**Comment by lauraykerr** [comment]
@georgehager Sounds good. On second glance, the "Thanks for choosing SimpleReport" seems a little out of place since it's not really functioning as a sign-off in this instance. If it's ok, would prefer to remove. 

Otherwise looks great to me, thank you!

**Comment by georgehager** [comment]
@lauraykerr how about this?

![Screen Shot 2022-11-22 at 8 51 06 AM](https://user-images.githubusercontent.com/89797785/203330865-c74e9c02-9b15-42f6-a981-dbc63208d103.png)

**Comment by lauraykerr** [comment]
Looks good to me, thank you @georgehager!

**Comment by georgehager** [comment]
@lauraykerr just one more question! In the event that upload fails for some reason (after validation and while the upload is processing), should we send an email to the user in that case? If so, what would you like that to look like?

**Comment by lauraykerr** [comment]
@georgehager With your recs from Slack conversation, suggest the following: 

> Greetings from SimpleReport, 
>
> Sorry, our system had a problem while adding your patients. Please try uploading the file again. If you still have issues, you can reach out to [support@simplereport.gov](url).
>
> [Button] Visit SimpleReport 
>
> [SimpleReport terms of service](url)

**Comment by georgehager** [comment]
@lauraykerr how does this look?

![Screen Shot 2022-11-22 at 4 33 48 PM](https://user-images.githubusercontent.com/89797785/203425693-8d86e310-be2a-4a40-8482-21b16e8bd40a.png)

**Comment by lauraykerr** [comment]
@georgehager Poifect, thanks!

**Comment by kenieh** [comment]
Sorry y'all, realize I never got back to this. LGTM!
